README.md
# [Tabular](https://github.com/rtmigo/tabular_dart#tabular)

Dart library for displaying tabular data in a visually appealing 
ASCII table format.

It converts a two-dimensional array of cells into a string with a text
table.

Tabular is specifically designed to create tables in the Markdown format that 
GitHub understands.

The library is inspired by python's [tabulate](https://pypi.org/project/tabulate/) 
and [pretty_table](https://pypi.org/project/prettytable/). 

----


```dart
import 'package:tabular/tabular.dart';

var data = [
    ['Season', '#', 'Name', 'Days', 'Sun'],
    ['Winter', 1, 'January', 31, 94],
    ['Winter', 2, 'February', 28, 123],
    ['Spring', 3, 'March', 31, 42],
    ['Spring', 4, 'April', 30, 243],
    ['Spring', 5, 'May', 31, 5523],
    ['Summer', 6, 'June', 30, 11251],
    ['Summer', 7, 'July', 31, 17451],
    ['Summer', 8, 'August', 31, 18707],
    ['Autumn', 9, 'September', 30, 7025],
    ['Autumn', 10, 'October', 31, 5041],
    ['Autumn', 11, 'November', 30, 2302],
    ['Winter', 12, 'December', 31, 258],
];

var string = tabular(data);

print(string);
```

```text
Season |  # | Name      | Days |   Sun
-------|----|-----------|------|------
Winter |  1 | January   |   31 |    94
Winter |  2 | February  |   28 |   123
Spring |  3 | March     |   31 |    42
Spring |  4 | April     |   30 |   243
Spring |  5 | May       |   31 |  5523
Summer |  6 | June      |   30 | 11251
Summer |  7 | July      |   31 | 17451
Summer |  8 | August    |   31 | 18707
Autumn |  9 | September |   30 |  7025
Autumn | 10 | October   |   31 |  5041
Autumn | 11 | November  |   30 |  2302
Winter | 12 | December  |   31 |   258
```


# Formatting

## Modify alignment

```dart
tabular(data, align: {'Name': Side.center, 'Sun': Side.start})
```

```text
Season |  # |    Name   | Days | Sun  
-------|----|-----------|------|------
Winter |  1 |  January  |   31 | 94   
Winter |  2 |  February |   28 | 123  
Spring |  3 |   March   |   31 | 42   
Spring |  4 |   April   |   30 | 243  
Spring |  5 |    May    |   31 | 5523 
Summer |  6 |    June   |   30 | 11251
Summer |  7 |   July    |   31 | 17451
Summer |  8 |   August  |   31 | 18707
Autumn |  9 | September |   30 | 7025 
Autumn | 10 |  October  |   31 | 5041 
Autumn | 11 | November  |   30 | 2302 
Winter | 12 |  December |   31 | 258  
```

## Add markdown ':' alignment marks

```dart
tabular(data, markdownAlign: true)
```

```text
Season |  # | Name      | Days |   Sun
-------|---:|-----------|-----:|-----:
Winter |  1 | January   |   31 |    94
Winter |  2 | February  |   28 |   123
Spring |  3 | March     |   31 |    42
Spring |  4 | April     |   30 |   243
Spring |  5 | May       |   31 |  5523
Summer |  6 | June      |   30 | 11251
Summer |  7 | July      |   31 | 17451
Summer |  8 | August    |   31 | 18707
Autumn |  9 | September |   30 |  7025
Autumn | 10 | October   |   31 |  5041
Autumn | 11 | November  |   30 |  2302
Winter | 12 | December  |   31 |   258
```

## Add horizontal dividers

```dart
tabular(data, rowDividers: [1, 3, 6, 9, 12])
```

```text
Season |  # | Name      | Days |   Sun
-------|----|-----------|------|------
Winter |  1 | January   |   31 |    94
Winter |  2 | February  |   28 |   123
-------|----|-----------|------|------
Spring |  3 | March     |   31 |    42
Spring |  4 | April     |   30 |   243
Spring |  5 | May       |   31 |  5523
-------|----|-----------|------|------
Summer |  6 | June      |   30 | 11251
Summer |  7 | July      |   31 | 17451
Summer |  8 | August    |   31 | 18707
-------|----|-----------|------|------
Autumn |  9 | September |   30 |  7025
Autumn | 10 | October   |   31 |  5041
Autumn | 11 | November  |   30 |  2302
-------|----|-----------|------|------
Winter | 12 | December  |   31 |   258
```

Please be aware that dividers can make the table not conform to the 
Markdown standard.

## Add outer border

```dart
tabular(data, border: Border.vertical);
```

```text
| Season |  # | Name      | Days |   Sun |
|--------|----|-----------|------|-------|
| Winter |  1 | January   |   31 |    94 |
| Winter |  2 | February  |   28 |   123 |
| Spring |  3 | March     |   31 |    42 |
| Spring |  4 | April     |   30 |   243 |
| Spring |  5 | May       |   31 |  5523 |
| Summer |  6 | June      |   30 | 11251 |
| Summer |  7 | July      |   31 | 17451 |
| Summer |  8 | August    |   31 | 18707 |
| Autumn |  9 | September |   30 |  7025 |
| Autumn | 10 | October   |   31 |  5041 |
| Autumn | 11 | November  |   30 |  2302 |
| Winter | 12 | December  |   31 |   258 |
```

In the same way, you can add `Border.horizontal` or `Border.all`. But it can 
make the table not conform to the Markdown standard.

## Switch border style

``` dart
tabular(data, style: Style.mysql, border: Border.all)
```

``` text
+--------+----+-----------+------+-------+
| Season |  # | Name      | Days |   Sun |
+--------+----+-----------+------+-------+
| Winter |  1 | January   |   31 |    94 |
| Winter |  2 | February  |   28 |   123 |
| Spring |  3 | March     |   31 |    42 |
| Spring |  4 | April     |   30 |   243 |
| Spring |  5 | May       |   31 |  5523 |
| Summer |  6 | June      |   30 | 11251 |
| Summer |  7 | July      |   31 | 17451 |
| Summer |  8 | August    |   31 | 18707 |
| Autumn |  9 | September |   30 |  7025 |
| Autumn | 10 | October   |   31 |  5041 |
| Autumn | 11 | November  |   30 |  2302 |
| Winter | 12 | December  |   31 |   258 |
+--------+----+-----------+------+-------+
```


## Custom column formatting

``` dart
tabular(data, 
    format: {
        'Name': (value) => value.toUpperCase(), 
        'Sun': (value) => '+ '+value.toString() })
```

``` text
Season |  # | Name      | Days |     Sun
-------|----|-----------|------|--------
Winter |  1 | JANUARY   |   31 |    + 94
Winter |  2 | FEBRUARY  |   28 |   + 123
Spring |  3 | MARCH     |   31 |    + 42
Spring |  4 | APRIL     |   30 |   + 243
Spring |  5 | MAY       |   31 |  + 5523
Summer |  6 | JUNE      |   30 | + 11251
Summer |  7 | JULY      |   31 | + 17451
Summer |  8 | AUGUST    |   31 | + 18707
Autumn |  9 | SEPTEMBER |   30 |  + 7025
Autumn | 10 | OCTOBER   |   31 |  + 5041
Autumn | 11 | NOVEMBER  |   30 |  + 2302
Winter | 12 | DECEMBER  |   31 |   + 258
```



# Sorting

## Sort alphabetically by column 0

``` dart
tabular(data, sort: [Sort(0)]) // 0 is 'Season'
```

``` text
Season |  # | Name      | Days |   Sun
-------|----|-----------|------|------
Autumn |  9 | September |   30 |  7025
Autumn | 10 | October   |   31 |  5041
Autumn | 11 | November  |   30 |  2302
Spring |  3 | March     |   31 |    42
Spring |  4 | April     |   30 |   243
Spring |  5 | May       |   31 |  5523
Summer |  6 | June      |   30 | 11251
Summer |  7 | July      |   31 | 17451
Summer |  8 | August    |   31 | 18707
Winter |  1 | January   |   31 |    94
Winter |  2 | February  |   28 |   123
Winter | 12 | December  |   31 |   258
```

## Sort by 'Days' descending, and then by 'Sun' ascending

``` dart
tabular(data, sort: [Sort('Days', ascending: false), Sort('Sun')])
```

``` text
Season |  # | Name      | Days |   Sun
-------|----|-----------|------|------
Spring |  3 | March     |   31 |    42
Winter |  1 | January   |   31 |    94
Winter | 12 | December  |   31 |   258
Autumn | 10 | October   |   31 |  5041
Spring |  5 | May       |   31 |  5523
Summer |  7 | July      |   31 | 17451
Summer |  8 | August    |   31 | 18707
Spring |  4 | April     |   30 |   243
Autumn | 11 | November  |   30 |  2302
Autumn |  9 | September |   30 |  7025
Summer |  6 | June      |   30 | 11251
Winter |  2 | February  |   28 |   123
```

## Sort with custom comparison

Ordering the rows by the length of the month name.

``` dart
tabular(data, 
    sort: [Sort('Name', compare: (a, b) => a.length.compareTo(b.length))])
```

``` text
Season |  # | Name      | Days |   Sun
-------|----|-----------|------|------
Spring |  5 | May       |   31 |  5523
Summer |  6 | June      |   30 | 11251
Summer |  7 | July      |   31 | 17451
Spring |  3 | March     |   31 |    42
Spring |  4 | April     |   30 |   243
Summer |  8 | August    |   31 | 18707
Winter |  1 | January   |   31 |    94
Autumn | 10 | October   |   31 |  5041
Winter |  2 | February  |   28 |   123
Autumn | 11 | November  |   30 |  2302
Winter | 12 | December  |   31 |   258
Autumn |  9 | September |   30 |  7025
```

CHANGELOG.md
# 0.6.3+1

- Added `rowDividers` argument

# 0.5.1

- Added `style` argument to switch between Markdown and MySQL
- Added `border` argument, allowing vertical and horizontal outer borders
- `outerBorder` argument is deprecated 

# 0.4.1+1

- Added the option of custom sorting with callbacks

# 0.3.2-beta

- Documentation additions

# 0.3.1-beta

- Modified the way to align cells to the center

# 0.3.0-beta

- Documentation additions

# 0.2.2-alpha

- Added `format` argument
- Added `align` argument

# 0.1.0-alpha

- Initial release

test/data.dart
String mini(String s) {
  final strippedRows = s.split('\n').map((s) => s.trim()).toList();
  while (strippedRows.isNotEmpty && strippedRows.first.isEmpty) {
    strippedRows.removeAt(0);
  }

  while (strippedRows.isNotEmpty && strippedRows.last.isEmpty) {
    strippedRows.removeLast();
  }

  return strippedRows.join('\n');
}

final months = [
  ['Winter', 'Spring', 'Summer', 'Autumn'],
  ['December', 'March', 'June', 'September'],
  ['January', 'April', 'July', 'October'],
  ['February', 'May', 'August', 'November']
];

final numbers = [
  ['First', 'Second', 'Third'],
  [85,    10,    27],
  [512, 2348, 12037],
  [5,     10,     3],
  [512,   -1,   037],
  [1,     10,   100],
];

final numbersWithNulls = [
  ['First', 'Second', 'Third'],
  [85,    10,    27],
  [512, 2348, null],
  [null,     10,    3],
  [512,   -1,   null],
  [1,     10,   100],
];




test/common.dart
// SPDX-FileCopyrightText: (c) 2021 Artёm IG <github.com/rtmigo>
// SPDX-License-Identifier: MIT

String trimLR(String s, {onlyRight = false}) {
  final strippedRows = s.split('\n').map((s) => onlyRight ? s.trimRight() : s.trim()).toList();
  while (strippedRows.isNotEmpty && strippedRows.first.isEmpty) {
    strippedRows.removeAt(0);
  }

  while (strippedRows.isNotEmpty && strippedRows.last.isEmpty) {
    strippedRows.removeLast();
  }

  return strippedRows.join('\n');
}

String trimR(String s) {
  return trimLR(s, onlyRight: true);
}

List<String> removeCommonBlanksAtLeft(List<String> lines) {
  while (!lines.any((element) => element.isEmpty)) {
    // while there are no empty lines (so they all have a first char)
    final firstCharSet = lines.map((str) => str[0]).toSet();
    if (firstCharSet.length == 1 && ['\t', ' '].contains(firstCharSet.first)) {
      // all the lines have exactly the same first char, and it's a blank.
      // Recreating the lines list with first char removed
      lines = lines.map((str) => str.substring(1)).toList();
      continue; // try again, maybe there are more blanks to remove
    }
    break; // there are no common blanks anymore
  }
  return lines;
}

String testTrim(String s) {
  var lines = s.split('\n');

  // removing empty lines at the beginning of the list
  while (lines.isNotEmpty && lines.first.trim().isEmpty) {
    lines.removeAt(0);
  }

  // removing empty lines at the end of the list
  while (lines.isNotEmpty && lines.last.trim().isEmpty) {
    lines.removeLast();
  }

  lines = removeCommonBlanksAtLeft(lines);

  return lines.join('\n');
}

extension TestExt on String {
  String spaceToDot() {
    return this.replaceAll(' ', '.');
  }

  String noSpaces() {
    return this.replaceAll('\t', '').replaceAll(' ', '');
  }
}

var sundata = [
  ['Season', '#', 'Name', 'Days', 'Sun'],
  ['Winter', 1, 'January', 31, 94],
  ['Winter', 2, 'February', 28, 123],
  ['Spring', 3, 'March', 31, 42],
  ['Spring', 4, 'April', 30, 243],
  ['Spring', 5, 'May', 31, 5523],
  ['Summer', 6, 'June', 30, 11251],
  ['Summer', 7, 'July', 31, 17451],
  ['Summer', 8, 'August', 31, 18707],
  ['Autumn', 9, 'September', 30, 7025],
  ['Autumn', 10, 'October', 31, 5041],
  ['Autumn', 11, 'November', 30, 2302],
  ['Winter', 12, 'December', 31, 258],
];

final zdata = [
  ['Continent', 'Country', 'Islands', 'Population', 'Per island'],
  ['Europe', 'Norway', 55000, 5421241, 98.6],
  ['North America', 'Canada', 52455, 37742154, 719.5],
  ['Asia', 'Indonesia', 17508, 273523615, 15622.8],
  ['Europe', 'Finland', 188000, 5540720, 29.5],
  ['Asia', 'Japan', 6853, 126476461, 18455.6],
  ['Europe', 'Sweden', 221800, 10099265, 45.5],
];

final months = [
  ['Winter', 'Spring', 'Summer', 'Autumn'],
  ['December', 'March', 'June', 'September'],
  ['January', 'April', 'July', 'October'],
  ['February', 'May', 'August', 'November']
];

final numbers = [
  ['First', 'Second', 'Third'],
  [85, 10, 27],
  [512, 2348, 12037],
  [5, 10, 3],
  [512, -1, 037],
  [1, 10, 100],
];

final numbersWithNulls = [
  ['First', 'Second', 'Third'],
  [85, 10, 27],
  [512, 2348, null],
  [null, 10, 3],
  [512, -1, null],
  [1, 10, 100],
];


test/samples.dart
import 'package:tabular/tabular.dart';

void main() {
  final zdata = [
    ['Continent', 'Country', 'Islands', 'Population', 'Per island'],
    ['Europe', 'Norway', 55000, 5421241, 98.6],
    ['North America', 'Canada', 52455, 37742154, 719.5],
    ['Asia', 'Indonesia', 17508, 273523615, 15622.8],
    ['Europe', 'Finland', 188000, 5540720, 29.5],
    ['Asia', 'Japan', 6853, 126476461, 18455.6],
    ['Europe', 'Sweden', 221800, 10099265, 45.5],
  ];

  var data = [
    ['Season', '#', 'Name', 'Days', 'Sun'],
    ['Winter', 1, 'January', 31, 94],
    ['Winter', 2, 'February', 28, 123],
    ['Spring', 3, 'March', 31, 42],
    ['Spring', 4, 'April', 30, 243],
    ['Spring', 5, 'May', 31, 5523],
    ['Summer', 6, 'June', 30, 11251],
    ['Summer', 7, 'July', 31, 17451],
    ['Summer', 8, 'August', 31, 18707],
    ['Autumn', 9, 'September', 30, 7025],
    ['Autumn', 10, 'October', 31, 5041],
    ['Autumn', 11, 'November', 30, 2302],
    ['Winter', 12, 'December', 31, 258],
  ];

  print(tabular(data));
  print('');

  print(tabular(data, markdownAlign: true));
  print('');
  print(tabular(data, sort: [Sort(0)]));
  print('');
  print(tabular(data, sort: [Sort('Days', false), Sort('Sun')]));
  print('');

  print(tabular(data, outerBorder: true));

  // print('');
  //
  // print(tabular(data, sort: [Sort('Per island')]));
  // print('');
  // print(tabular(data, sort: [Sort('Per island', false)]));
  // print('');
  // print(tabular(data, sort: [Sort('Continent'), Sort('Country')]));

}


pubpub.sh
#!/bin/bash
set -e && cd "${0%/*}"

# creates a copy of the project in temporary directory
# and prepares the copy to be published

temp_pub_dir=$(mktemp -d -t pub-XXXXXXX)

echo "$temp_pub_dir"

rsync -Rrv ./ "$temp_pub_dir" \
  --exclude=".git" \
  --exclude="pubpub.sh" \
  --exclude="todo.txt" \
  --exclude="benchmark/" \
  --exclude="reference/" \
  --exclude="README.md" \
  --exclude=".github" \
  --exclude="labuda/" \
  --exclude="draft/" \
  --exclude="analysis_options.yaml" \
  --exclude="experiments"

# replacing \n by \r
old_readme=$(cat README.md | tr '\n' '\r')
# removing everything before "\n# ", the first header
new_readme=$(echo "$old_readme" | perl -p0e 's|^.*?\r# |# \1|')
# replacing \r by \n
new_readme=$(echo "$new_readme" |  tr '\r' '\n')

echo "$new_readme" > "$temp_pub_dir/README.md"

cd "$temp_pub_dir"
dartfmt -w .
dart pub publish

lib/src/inner.dart
// SPDX-FileCopyrightText: (c) 2021 Artёm IG <github.com/rtmigo>
// SPDX-License-Identifier: MIT

import 'dart:math';

import 'package:quiver/iterables.dart' show enumerate;

/// Sets the cell alignment. In the case of horizontal alignment, [Side.start] means left,
/// [Side.end] means right.
enum Side { start, end, center }

/// Describes the rules by which an individual column should be aligned.
class Align {
  Align(this.column, this.side);

  /// Identifies the column that we are sorting. This can be either [int],
  /// which specifies the column index, or [String], which specifies
  /// the column name.
  final dynamic column;

  /// Specifies which way each column cell should be shifted after converting
  /// the cell value to [String].
  final Side side;
}

enum Border {
  /// Don't add a border outside the table
  none,

  /// Add a border to the top and bottom of the table
  vertical,

  /// Add a border to the left and right of the table.
  horizontal,

  /// Add borders on all sides of the table: top, bottom, left, right
  all
}

// https://ozh.github.io/ascii-tables/

/// The border style.
enum Style { markdown, mysql }

typedef CompareCells = int Function(dynamic a, dynamic b);

/// Describes the rules by which an individual column should be sorted.
class Sort {
  Sort(this.column, {this.ascending = true, this.emptyFirst = false, this.compare});

  /// Identifies the column that we are sorting. This can be either [int],
  /// which specifies the column index, or [String], which specifies
  /// the column name.
  dynamic column;

  /// Defines the sort order of cells that have values. That is, cells that are
  /// not null or not an empty string.
  final bool ascending;

  /// Specifies whether the empty values should be placed at the beginning
  /// or at the end of the sorted column.
  final bool emptyFirst;

  final CompareCells? compare;
}

class Aligner {
  String alignText(String text, int targetWidth, Side align) {
    switch (align) {
      case Side.start:
        return text.padRight(targetWidth);
      case Side.end:
        return text.padLeft(targetWidth);
      case Side.center:
        return alignTextCenter(text, targetWidth);
    }
  }

  bool centerOdd = true; // that's why we need the object

  String alignTextCenter(String text, int targetWidth) {
    final halfD = (targetWidth - text.length) / 2;

    final half = centerOdd ? halfD.ceil() : halfD.floor(); //>> 1;
    centerOdd = !centerOdd;
    text = text.padLeft(text.length + half);
    text = text.padRight(targetWidth);
    return text;
  }
}

/// Specifies how to convert the source value of the cell to a [String].
typedef FormatCell = String Function(dynamic value);

class Cell implements Comparable<Cell> {
  Cell(this.rawCell, {this.nullReplacement = ''});

  final dynamic rawCell;
  final String nullReplacement;
  FormatCell? format;

  @override
  String toString() {
    if (rawCell == null) {
      return nullReplacement;
    } else {
      return rawCell.toString();
    }
  }

  String toFinalString() {
    // todo cache
    if (this.format != null) {
      return this.format!(this.rawCell);
    }
    return this.toString();
  }

  bool get isEmpty {
    return this.rawCell == null || this.rawCell.toString().isEmpty;
  }

  bool isNumber() {
    if (rawCell == null || rawCell == '' || rawCell is num) {
      return true;
    }

    final s = rawCell.toString();

    return int.tryParse(s) != null || double.tryParse(s) != null;
  }

  Side guessAlign() {
    return isNumber() ? Side.end : Side.start;
  }

  num? tryGetNum() {
    // todo cache
    if (this.rawCell is num) {
      return this.rawCell;
    }
    return int.tryParse(this.rawCell.toString()) ?? double.tryParse(this.rawCell.toString());
  }

  @override
  int compareTo(Cell other) {
    num? thisNum = this.tryGetNum();
    if (thisNum != null) {
      num? otherNum = other.tryGetNum();
      if (otherNum != null) {
        return thisNum.compareTo(otherNum);
      }
    }
    // todo null
    return this.toString().compareTo(other.toString());
  }
}

class CellsColumn {
  CellsColumn(this.matrix, this.index) {
    // if (cells.length <= 1) {
    //   throw ArgumentError.value(cells.length, 'cells.length');
    // }
  }

  final int index;
  final CellsMatrix matrix;

  Iterable<Cell> get cells sync* {
    for (var i = 0; i < matrix.rows.length; ++i) {
      yield this.matrix.rows[i][this.index];
    }
  }

  int get textWidth {

    int result = this.cells.map<int>((cell) => cell.toFinalString().length).reduce(max);

    if (result<=0) {
      result = 1;
    }
    return result;
    // todo cache
    //return this.cells.map<int>((cell) => cell.toFinalString().length).reduce(max);
  }

  Side guessAlign() {
    Side? result;

    for (final cell in this.cells.skip(1)) {
      if (result == null) {
        result = cell.guessAlign();
      } else if (cell.guessAlign() != result) {
        return Side.start; // default
      }
    }

    result ??= Side.start;
    return result;
  }
}

class CellsMatrix {
  CellsMatrix(List<List<dynamic>> rawRows, Map<dynamic, FormatCell>? format) {
    // if (rawRows.length <= 1) {
    //   throw ArgumentError.value(rawRows.length, 'rawRows.length',
    //       'Must contain at least two items: the header and the first row.');
    // }

    // determining the maximum count of cells in each row
    final columnsCount = rawRows.map<int>((r) => r.length).reduce(max);

    if (columnsCount <= 0) {
      throw ArgumentError('rawRows contains zero columns.');
    }

    // creating [rows] field
    for (final srcRow in rawRows) {
      // copying raw cell data into list on Cells
      final newRow = srcRow.map((raw) => Cell(raw)).toList();
      // adding empty cells to the end of the row, if needed
      while (newRow.length < columnsCount) {
        newRow.add(Cell(null));
      }
      this.rows.add(newRow);
    }

    // [rows] now have equal length, and the [header] returns row 0
    assert(this.rows.every((row) => row.length == this.header.length));

    // creating the columns list
    for (var i = 0; i < columnsCount; ++i) {
      this.columns.add(CellsColumn(this, i));
    }

    // assigning [format] to cells if its set in the argument
    assert(this.rows.isNotEmpty); // can use columnIndex
    assert(this.columns.isNotEmpty); // can iterate cells by column
    if (format != null) {
      for (final me in format.entries) {
        final iCol = this.columnIndex(me.key);
        for (final cell in this.columns[iCol].cells.skip(1)) {
          cell.format = me.value;
        }
      }
    }
  }

  List<Cell> get header => rows[0];

  List<List<Cell>> rows = <List<Cell>>[];

  Iterable<Cell> iterCellsByColumn(int columnIdx) sync* {
    for (var i = 0; i < rows.length; ++i) {
      yield rows[i][columnIdx];
    }
  }

  List<CellsColumn> columns = <CellsColumn>[];

  int columnIndex(dynamic col) {
    if (col is int) {
      return col;
    }
    String colStr = col.toString();
    for (var me in enumerate(header)) {
      if (me.value.toString() == colStr) {
        return me.index;
      }
    }
    throw ArgumentError.value(col, 'col', 'Column not found.');
  }

  void sortBy(List<Sort> rules) {
    if (this.columns.isEmpty) {
      return;
    }

    final indexedRules =
        rules.map((e) => MapEntry<int, Sort>(this.columnIndex(e.column), e)).toList();

    // replacing header with empty placeholder
    final temporaryRemovedHeader = this.rows[0];
    this.rows[0] = [];
    // since this.columnsCount>0, only a placeholder can be an empty row

    try {
      this.rows.sort((rowA, rowB) {
        if (rowA.isEmpty) {
          // header placeholder
          return -1;
        }

        if (rowB.isEmpty) {
          // header placeholder
          return 1;
        }

        for (final entry in indexedRules) {
          int columnIndex = entry.key;
          Sort rule = entry.value;

          var cell1 = rowA[columnIndex];
          var cell2 = rowB[columnIndex];

          if (rule.compare != null) {
            int customComparison = rule.compare!(cell1.rawCell, cell2.rawCell);
            return rule.ascending ? customComparison : -customComparison;
          }

          if (cell1.isEmpty || cell2.isEmpty) {
            if (cell1.isEmpty && cell2.isEmpty) {
              continue;
            }
            return cell1.isEmpty ? (rule.emptyFirst ? -1 : 1) : (rule.emptyFirst ? 1 : -1);
          }

          final A_IS_SMALLER = rule.ascending ? -1 : 1;
          final A_IS_LARGER = rule.ascending ? 1 : -1;

          int cmp = cell1.compareTo(cell2);
          if (cmp == 0) {
            continue;
          }
          if (cmp > 0) {
            return A_IS_LARGER;
          }
          return A_IS_SMALLER;
        }
        ;
        return 0;
      });
    } finally {
      this.rows[0] = temporaryRemovedHeader;
    }
  }
}

extension ListExt<T> on List<T> {
  void fillAndSet(int index, T value, T empty) {
    while (this.length < index) {
      this.add(empty);
    }
    if (this.length > index) {
      this[index] = value;
    } else {
      this.add(value);
    }
  }

  T tryGet(int index, T empty) {
    return (this.length > index) ? this[index] : empty;
  }
}

/// Transforms a map with optional alignment rules to a list will all elements set.
/// @returns A `result` list such as `result[columnIndex]` is the alignment for the column.
List<Side> createColToAlign<T>(CellsMatrix matrix, Map<dynamic, Side>? align) {
  List<Side?> colToAlignNullable = <Side?>[];
  if (align != null) {
    for (final me in align.entries) {
      int iCol = matrix.columnIndex(me.key);
      colToAlignNullable.fillAndSet(iCol, me.value, null);
    }
  }
  for (var iCol = 0; iCol < matrix.columns.length; iCol++) {
    if (colToAlignNullable.tryGet(iCol, null) == null) {
      colToAlignNullable.fillAndSet(iCol, matrix.columns[iCol].guessAlign(), null);
    }
  }

  List<Side> colToAlign = colToAlignNullable.map((e) => e!).toList();

  return colToAlign;
}

/// Converts a set of cells defined by a two-dimensional list to a Markdown formatted ASCII table.
/// Returns a string with the table that is ready for printing.
///
/// [rows] is the source data for the table. The first of the lists will be the header, the
/// others - ordinary rows. The values can be [String]s, [num]s, [null], or any objects. Either
/// way, they end up being converted to strings.
///
/// [align] specifies which way all cells in a particular column should be aligned. The keys
/// of this [Map] can be of type [int] then they denote the index of the column, or of type
/// [String] - then they denote the name of the column.
///
/// [format] specifies how to convert each cell of a particular column to a row.
/// This conversion occurs after sorting, but before alignment. The keys of this [Map] can be
/// of type [int] then they denote the index of the column, or of type [String] - then they
/// denote the name of the column.
///
/// [sort] determines the sorting order. Sorting can take place in several columns at once.
/// Priority will be given to the ones at the beginning of the list.
///
/// [markdownAlign] determines whether to add the ':' characters to the delimiter under
/// the header. These symbols tell services like GitHub which way to align the columns after
/// converting the table to HTML.
///
/// [border] determines whether an outer border should be added to the table and what parts
/// it should consist of.
///
/// [rowDividers] contains the indices of the [rows], which must be preceded by a horizontal
/// divider. By default, there is only index 1, which corresponds to the divider between
/// the header and the body of the table. `rowDividers: null` will cause dividers to be added
/// between all rows.
String tabular(
    List<List<dynamic>> rows,
    { Map<dynamic, Side>? align,
      Map<dynamic, FormatCell>? format,
      List<Sort>? sort,
      markdownAlign = false,
      Border border = Border.none,
      Style style = Style.markdown,
      List<int>? rowDividers = const [1],

      @Deprecated('Use border=Border.vertical argument') // since 2021-04-28
      outerBorder = false
    }) {

  if (rows.isEmpty) {
    throw ArgumentError.value(rows, 'rows', 'Must not be empty');
  }

  if (outerBorder && border==Border.none) {
    border = Border.vertical;
  }

  bool borderV = border == Border.vertical || border == Border.all;
  bool borderH = border == Border.horizontal || border == Border.all;

  final matrix = CellsMatrix(rows, format);

  if (sort != null) {
    matrix.sortBy(sort);
  }

  List<Side> colToAlign = createColToAlign(matrix, align);

  String cross;
  switch (style) {
    case Style.markdown:
      cross = '|';
      break;
    case Style.mysql:
      cross = '+';
      break;
  }

  String bar = '';
  if (borderV) {
    bar += cross;
  }
  for (int i = 0; i < matrix.columns.length; ++i) {
    final align = markdownAlign ? colToAlign[i] : null;
    int width = matrix.columns[i].textWidth;

    // if (width<=0) {
    //   width = 1;
    // }


    final bool isSingleColumn = matrix.columns.length==1;
    final bool isFirstColumn = i == 0;
    final bool isLastColumn = i == matrix.columns.length - 1;

    int extra = 0;
    if (!borderV) {
      if (isSingleColumn) {
        extra = -2;
      }
      else if (isFirstColumn || isLastColumn) {
        extra = -1;
      }
    }

    //int extra = ((isFirstColumn || isLastColumn) && !borderV) ? -1 : 0;


    switch (align) {
      case null:
      case Side.start:
        //print("A $width $extra");
        bar += ('-' * (width + 2 + extra));
        break;
      case Side.end:
        //print("B");
        bar += ('-' * (width + 1 + extra) + ':');
        break;
      case Side.center:
        //print("C");
        bar += (':' + '-' * width + ':');
    }

    if (borderV || !isLastColumn) {
      bar += cross;
    }

    //print("width $width $bar");
  }

  String dashBar() => bar; // '+'+('-'*(bar.length-2))+'+';

  final formattedRows = <String>[];

  if (borderH) {
    formattedRows.add(dashBar());
  }

  final aligner = Aligner();

  var iRow = -1;
  for (var row in matrix.rows) {
    iRow++;


    if (iRow>0 && (rowDividers==null || rowDividers.contains(iRow))) {
      formattedRows.add(bar);
    }

    var formatted = '';

    var iCol = -1;

    for (final me in enumerate(row)) {
      final cell = me.value;
      if (me.index == 0) {
        if (borderV) {
          formatted += '| ';
        }
      } else {
        formatted += ' | ';
      }
      iCol++;
      formatted +=
          aligner.alignText(cell.toFinalString(), matrix.columns[iCol].textWidth, colToAlign[iCol]);
    }

    if (borderV) {
      formatted += ' |';
    }

    formattedRows.add(formatted);
  }

  if (borderH) {
    formattedRows.add(dashBar());
  }

  return formattedRows.join('\n');
}


example/example.dart
import 'package:tabular/tabular.dart';

void main() {
  var data = [
    ['Season', '#', 'Name', 'Days', 'Sun'],
    ['Winter', 1, 'January', 31, 94],
    ['Winter', 2, 'February', 28, 123],
    ['Spring', 3, 'March', 31, 42],
    ['Spring', 4, 'April', 30, 243],
    ['Spring', 5, 'May', 31, 5523],
    ['Summer', 6, 'June', 30, 11251],
    ['Summer', 7, 'July', 31, 17451],
    ['Summer', 8, 'August', 31, 18707],
    ['Autumn', 9, 'September', 30, 7025],
    ['Autumn', 10, 'October', 31, 5041],
    ['Autumn', 11, 'November', 30, 2302],
    ['Winter', 12, 'December', 31, 258],
  ];

  String title(s) => '\n\n$s\n';

  print(title('JUST TABLE'));
  print(tabular(data));

  print(title('WITH VERTICAL BORDER'));
  print(tabular(data, border: Border.vertical));

  print(title('MYSQL STYLE'));
  print(tabular(data, style: Style.mysql, border: Border.all));

  print(title('WITH ADDITIONAL DIVIDERS'));
  print(tabular(data, rowDividers: [1, 3, 6, 9, 12]));


  print(title('WITH MARKDOWN ":"'));
  print(tabular(data, markdownAlign: true));

  print(title('WITH MODIFIED ALIGNMENT'));
  print(tabular(data, align: {'Name': Side.center, 'Sun': Side.start}));

  print(title('WITH CUSTOM FORMATTING'));
  print(tabular(data,
      format: {'Name': (value) => value.toUpperCase(), 'Sun': (value) => '+ ' + value.toString()}));

  print(title('SORTED BY FIRST COLUMN'));
  print(tabular(data, sort: [Sort(0)]));

  print(title('SORTED BY TWO COLUMNS "DAYS" AND "SUN"'));
  print(tabular(data, sort: [Sort('Days', ascending: false), Sort('Sun')]));

  print(title('SORTED WITH A CUSTOM COMPARISON'));
  print(tabular(data, sort: [Sort('Name', compare: (a, b) => a.length.compareTo(b.length))]));
}
